ResourceInfo Linking Panic in terraform-provider-vsphere

## Version

```
tfproviderlint v0.15.0-dev
```

## Steps to Reproduce

```console
$ git clone git@github.com:terraform-providers/terraform-provider-vsphere.git
$ cd terraform-provider-vsphere
$ tfproviderlint -R003 ./...
```

## Stacktrace

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1a198d0]

goroutine 4673 [running]:
github.com/bflad/tfproviderlint/helper/terraformtype/helper/schema.NewResourceInfo(0xc002b1c9c0, 0xc00325c820, 0xc003e53d01)
	/Users/bflad/src/github.com/bflad/tfproviderlint/helper/terraformtype/helper/schema/type_resource.go:62 +0x280
github.com/bflad/tfproviderlint/passes/helper/schema/resourceinfo.run.func1(0x2034e20, 0xc002b1c9c0)
	/Users/bflad/src/github.com/bflad/tfproviderlint/passes/helper/schema/resourceinfo/resourceinfo.go:37 +0x82
golang.org/x/tools/go/ast/inspector.(*Inspector).Preorder(0xc003e53a60, 0xc003938b08, 0x1, 0x1, 0xc003938b30)
	/Users/bflad/go/pkg/mod/golang.org/x/tools@v0.0.0-20200414001008-ae52e4b55789/go/ast/inspector/inspector.go:77 +0x9f
github.com/bflad/tfproviderlint/passes/helper/schema/resourceinfo.run(0xc0000f83c0, 0xc0000f83c0, 0x0, 0x0, 0x0)
	/Users/bflad/src/github.com/bflad/tfproviderlint/passes/helper/schema/resourceinfo/resourceinfo.go:30 +0xef
golang.org/x/tools/go/analysis/internal/checker.(*action).execOnce(0xc003ecf720)
	/Users/bflad/go/pkg/mod/golang.org/x/tools@v0.0.0-20200414001008-ae52e4b55789/go/analysis/internal/checker/checker.go:690 +0x83a
```

## References

- https://github.com/bflad/tfproviderlint/pull/129
- https://github.com/bflad/tfproviderlint/blob/0dd4d5871e9209bd6e6fd9bce58dde124dc9f206/helper/terraformtype/helper/schema/type_resource.go#L62

--- a/helper/terraformtype/helper/schema/testdata/type_resource/resource.go
+++ b/helper/terraformtype/helper/schema/testdata/type_resource/resource.go
@@ -28,10 +28,101 @@ var ResourceFuncs = []resourceFunc{
 					Type:     schema.TypeList,
 					Required: true,
 					MaxItems: 1,
-					Elem: &schema.Resource{
+					Elem:     &schema.Resource{},
+				},
+			},
+		}
+	},
+
+	// resource with nested schema which has nested schema
+	func() *schema.Resource {
+		return &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"test": {
+					Type:     schema.TypeMap,
+					Optional: true,
+					Elem: &schema.Schema{
+						Type: schema.TypeString,
 					},
 				},
 			},
 		}
 	},
+
+	// Ensure no panic with constant keys
+	func() *schema.Resource {
+		return &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				ConstSchemaKeyTest: {
+					Type:     schema.TypeString,
+					Required: true,
+				},
+			},
+		}
+	},
+
+	// Ensure no panic with variable keys
+	func() *schema.Resource {
+		return &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				VarSchemaKeyTest: {
+					Type:     schema.TypeString,
+					Required: true,
+				},
+			},
+		}
+	},
+
+	// Ensure no panic with function call keys
+	func() *schema.Resource {
+		return &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				FuncSchemaKey(): {
+					Type:     schema.TypeString,
+					Required: true,
+				},
+			},
+		}
+	},
+
+	// Ensure no panic with variable values
+	func() *schema.Resource {
+		return &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"test": VarSchemaValue,
+			},
+		}
+	},
+
+	// Ensure no panic with function call values
+	func() *schema.Resource {
+		return &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"test": FuncSchemaValue(),
+			},
+		}
+	},
+}
+
+const (
+	ConstSchemaKeyTest = "test"
+)
+
+var (
+	VarSchemaKeyTest = "test"
+	VarSchemaValue   = &schema.Schema{
+		Type:     schema.TypeString,
+		Required: true,
+	}
+)
+
+func FuncSchemaKey() string {
+	return "test"
+}
+
+func FuncSchemaValue() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeString,
+		Required: true,
+	}
 }

--- a/helper/terraformtype/helper/schema/type_resource.go
+++ b/helper/terraformtype/helper/schema/type_resource.go
@@ -57,11 +57,29 @@ func NewResourceInfo(cl *ast.CompositeLit, info *types.Info) *ResourceInfo {
 		result.Resource.Schema = map[string]*tfschema.Schema{}
 		if smap, ok := kvExpr.Value.(*ast.CompositeLit); ok {
 			for _, expr := range smap.Elts {
-				kvExpr := expr.(*ast.KeyValueExpr) // safe cast, map elements are always kv
-				keyPtr := astutils.ExprStringValue(kvExpr.Key)
-				key := *keyPtr // safe, since schema map key must be string
-				if cl, ok := kvExpr.Value.(*ast.CompositeLit); ok {
-					result.Resource.Schema[key] = NewSchemaInfo(cl, info).Schema
+				switch elt := expr.(type) {
+				case *ast.KeyValueExpr:
+					var key string
+
+					switch keyExpr := elt.Key.(type) {
+					case *ast.BasicLit:
+						keyPtr := astutils.ExprStringValue(keyExpr)
+
+						if keyPtr == nil {
+							continue
+						}
+
+						key = *keyPtr
+					}
+
+					if key == "" {
+						continue
+					}
+
+					switch valueExpr := elt.Value.(type) {
+					case *ast.CompositeLit:
+						result.Resource.Schema[key] = NewSchemaInfo(valueExpr, info).Schema
+					}
 				}
 			}
 		}

--- a/helper/terraformtype/helper/schema/type_resource_test.go
+++ b/helper/terraformtype/helper/schema/type_resource_test.go
@@ -1,13 +1,15 @@
 package schema
 
 import (
-	"github.com/bflad/tfproviderlint/helper/terraformtype/helper/schema/testdata/type_resource"
 	"go/ast"
-	"golang.org/x/tools/go/packages"
 	"reflect"
+	"strings"
 	"testing"
-)
 
+	"github.com/bflad/tfproviderlint/helper/terraformtype/helper/schema/testdata/type_resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"golang.org/x/tools/go/packages"
+)
 
 func TestNewResourceInfo_Resource(t *testing.T) {
 	// load the single test file
@@ -20,16 +22,24 @@ func TestNewResourceInfo_Resource(t *testing.T) {
 	resourceFuncLits := f.Decls[2].(*ast.GenDecl).Specs[0].(*ast.ValueSpec).Values[0].(*ast.CompositeLit).Elts
 
 	for i, resourceFuncLit := range resourceFuncLits {
-		expectResource := type_resource.ResourceFuncs[i]()
-		desc := ast.NewCommentMap(pkg.Fset, resourceFuncLit, f.Comments)[resourceFuncLit][0].List[0].Text
+		desc := ast.NewCommentMap(pkg.Fset, resourceFuncLit, f.Comments)[resourceFuncLit][i].List[0].Text
+
+		var expectResource *schema.Resource
+		if strings.Contains(desc, "no panic") {
+			expectResource = &schema.Resource{
+				Schema: map[string]*schema.Schema{},
+			}
+		} else {
+			expectResource = type_resource.ResourceFuncs[i]()
+		}
 
 		// new resource info
 		cl := resourceFuncLit.(*ast.FuncLit).Body.List[0].(*ast.ReturnStmt).Results[0].(*ast.UnaryExpr).X.(*ast.CompositeLit)
 		resourceInfo := NewResourceInfo(cl, pkg.TypesInfo)
 
 		// test
-		if !reflect.DeepEqual(expectResource, resourceInfo.Resource) {
-			t.Fatalf("test case: %s, resource not match", desc)
+		if !reflect.DeepEqual(*expectResource, *resourceInfo.Resource) {
+			t.Fatalf("test case: %s, resource not match\n\nexpected: %#v\n\ngot: %#v\n\n", desc, *expectResource, *resourceInfo.Resource)
 		}
 	}
 }

